File: Programmers/[131703] 2차원 동전 뒤집기/KingDonggyu/131703.js
function filpRow(coins, row) {
  for (let col = 0; col < coins[0].length; col++) {
    coins[row][col] = coins[row][col] ? 0 : 1;
  }
}

function flipColumn(coins, col) {
  for (let row = 0; row < coins.length; row++) {
    coins[row][col] = coins[row][col] ? 0 : 1;
  }
}

function solution(beginning, target) {
  const N = beginning.length;
  const M = beginning[0].length;
  const MAX = N * M + 1;

  let answer = MAX;
  const visited = Array(N + M).fill(false);

  function dfs(count) {
    if (answer <= count) {
      return;
    }

    for (let row = 0; row < N; row++) {
      for (let col = 0; col < M; col++) {
        if (beginning[row][col] === target[row][col]) {
          continue;
        }

        // 행 뒤집기
        if (!visited[row]) {
          visited[row] = true;
          filpRow(beginning, row);

          dfs(count + 1);

          visited[row] = false;
          filpRow(beginning, row);
        }

        // 열 뒤집기
        if (!visited[N + col]) {
          visited[N + col] = true;
          flipColumn(beginning, col);

          dfs(count + 1);

          visited[N + col] = false;
          flipColumn(beginning, col);
        }

        // 실패
        return;
      }
    }

    // 성공
    answer = count;
  }

  dfs(0);

  if (answer < MAX) {
    return answer;
  }

  return -1;
}

File: Programmers/[131703] 2차원 동전 뒤집기/ONE/131703.java
class Solution {
    private int minCount = Integer.MAX_VALUE;
    private int N;
    private int M;
    private boolean[] visitedRow;
    private boolean[] visitedColumn;
    private int[][] beginning;
    private int[][] target;

    public int solution(int[][] beginning, int[][] target) {
        this.N = beginning.length;
        this.M = beginning[0].length;
        this.visitedRow = new boolean[N];
        this.visitedColumn = new boolean[M];
        this.beginning = beginning;
        this.target = target;

        simulate(0);

        return minCount == Integer.MAX_VALUE ? -1 : minCount;
    }

    private void simulate(int count) {
        if (count > minCount) {
            return;
        }

        for (int row = 0; row < N; row++) {
            for (int col = 0; col < M; col++) {
                if (beginning[row][col] == target[row][col]) {
                    continue;
                }
                if (!visitedRow[row]) {
                    visitedRow[row] = true;
                    flipRow(row);

                    simulate(count + 1);

                    visitedRow[row] = false;
                    flipRow(row);
                }
                if (!visitedColumn[col]) {
                    visitedColumn[col] = true;
                    flipColumn(col);

                    simulate(count + 1);

                    visitedColumn[col] = false;
                    flipColumn(col);
                }
                return;
            }
        }
        minCount = count;
    }

    private void flipRow(int row) {
        for (int index = 0; index < M; index++) {
            beginning[row][index] = beginning[row][index] == 0 ? 1 : 0;
        }
    }

    private void flipColumn(int col) {
        for (int index = 0; index < N; index++) {
            beginning[index][col] = beginning[index][col] == 0 ? 1 : 0;
        }
    }
}